src/Chapter3.hs

```diff
@@ -1006,6 +1171,223 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+data Com = Com
+  { comHealth    :: Int
+  , comAttack    :: Int
+  } deriving (Show)
+
+class ComCl a where
+  getCom :: a -> Com
+  setCom :: a -> Com -> a
+
+getHealth  :: (ComCl a) => a -> Int
+getHealth a     = comHealth (getCom a)
+
+getAttack  :: (ComCl a) => a -> Int
+getAttack a     = comAttack (getCom a)
+
+addHealth  :: (ComCl a) => a -> Int -> a
+addHealth a dh  = setCom a ((getCom a) { comHealth = h2})
+  where
+    h2 = (getHealth a) + dh
+
+
+data MAction = MAttack | Runaway        deriving (Show, Eq)
+data KAction = KAttack | Potion | Spell deriving (Show, Eq)
```

Comment:
I separated the data types for Knight and Monster to detect avoid invalid Actions in compile time not in runtime based on your suggestion.